activerecord/lib/active_record/relation/query_attribute.rb

```diff
@@ -10,7 +10,7 @@ def initialize(...)
 
         # The query attribute value may be mutated before we actually "compile" the query.
         # To avoid that if the type uses a serializer we eagerly compute the value for database
-        if @type.serialized?
+        if @type.serialized? && !value_before_type_cast.is_a?(StatementCache::Substitute)
```

Comment:
I suspect we don't care about the second branch either, so maybe we can move that condition as it's own `if` above this one?